Add exports=named to rollup output config

BPScott opened the pull request:
Currently index.ts contains a single default export, and as `output.exports` is not specified it uses the value of "default". This causes problems with generating CommonJS output that is meant to be interchangeable with ESM output as detailed in https://rollupjs.org/configuration-options/#output-exports and https://www.typescriptlang.org/docs/handbook/modules/appendices/esm-cjs-interop.html. This also causes problems with typescript's `moduleResolution:node16` option as the types types published by this package do match the commonjs output.

Prior to this change the types in `index.d.ts` are `import { webpackStats } from './plugin'; export default webpackStats;` which is correct for the esm build output.

The compiled `index.cjs` file however contains `module.exports = webpackStats;`. The correct types for this this content would be `export = webpackStats`.


---

To make the commonjs output match the types that are generated, and have commonjs and esm behave equivalently, this PR adds `exports: "named"` to the rollup output options.

This has no effect on the esm output, but changes the commonjs output in index.cjs file as follows:

```diff
- module.exports = webpackStats;
+ exports.default = webpackStats;
``` 

---

This is a breaking change, as it means commonjs consumers will have to change how they consume the package:

```diff
- const webpackStatsPlugin = require('rollup-plugin-webpack-stats');
+ const webpackStatsPlugin = require('rollup-plugin-webpack-stats').default
```

---

My general approach to working around this problem is "I only ever use named exports, I totally avoid `export default`", so personally I'd also change `export default webpackStats` to `export {webpackStats}` but I figure this is a less invasive change.

vio commented:
@BPScott thank you for the pr!

Changing the export is a breaking change, and in the long run, we prefer to use default exports for ease of use. We are already working on fixing the support across all the export types(cjs, mjs, d.ts), will provide an update later this week.

How did you encounter the error? Did you update the plugin, or did you change the type of the vite/rollup config?

BPScott commented:
I encountered this error when I changed the tsconfig that runs over my `vite.config.ts` (which imports `rollup-plugin-webpack-stats`).

I changed `{"module": "esnext", "moduleResolution": "bundler"}` to `{"module": "nodenext", "moduleResolution": "node16"}` to make typescript use the node module resolution strategies for my node based tools.

When I made this change I began to get the following error on the line that did `webpackPlugin()`

```
This expression is not callable.
  Type 'typeof import("/Users/ben/projects/rollup-plugin-webpack-stats-typeerror/node_modules/rollup-plugin-webpack-stats/dist/index")' has no call signatures.
```

---

I've created a reproduction demo fro you at https://github.com/BPScott/rollup-plugin-webpack-stats-typeerror

Clone the repo, and then run `npm run type-check` to run tsc and see the errors.

vio commented:
@BPScott thank you for repo! 

We've been testing `2.1.5-beta.5` with improved esm/cjs support, and it works ok for our automated tests(rollup/vite/rolldown) and for some Vite projects.

I tried to use it on your repo and got the same errors. Also, I tested with `@rollup/plugin-html`, which has a similar ESM/CJS setup, and it fails when running `tsc` with the same error. The shape of the types looks correct, and the modules are exporting a function as default(esm) or module.exports(cjs), though at some point during the check, the module definitions are seen as objects `{ default: fn }`.

I am planning to have a deeper look soon, but in the meantime, if you notice any other problems with the beta or if you have new  information regarding your setup, please feel free to add it here: https://github.com/relative-ci/rollup-plugin-webpack-stats/issues/616

BPScott commented:
> The shape of the types looks correct, and the modules are exporting a function as default(esm) or module.exports(cjs), though at some point during the check, the module definitions are seen as objects { default: fn }.

This is problematic and is what I was trying to point out, the shape of the types are incorrect for the runtime code you're producing.

The correct types for the esm code `export default function foo() {}` is `export default function foo(): void;`. What exists in your package is correct, the esm runtime code and types line up.

While the correct types for the cjs code `module.exports = function foo() {}` is `export =  function foo(): void;`. What exists in your package is incorrect, the commonjs runtime code and types do not match.


At the moment you're trying to use a single dts file for these two pieces of runtime code and that does not work as those two pieces of runtime code require different types definitions  per https://arethetypeswrong.github.io/?p=rollup-plugin-webpack-stats%402.1.4 and https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/main/docs/problems/FalseExportDefault.md.



The change in this PR attempts to resolve this problem by changing the commonjs code into a format that works with the single dts file you are generating.
